mig-agent/stats.go

```diff
@@ -0,0 +1,73 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Contributor: Aaron Meihm ameihm@mozilla.com [:alm]
+
+package main
+
+// Contains functions related to agent statistics. Currently this data is not
+// persistent and will be lost of the agent process is restarted, this could be
+// adapted to make use of local storage as needed.
+//
+// The information here can be presented by querying the agent on its status
+// socket.
+
+import (
+	"mig.ninja/mig"
+	"sync"
+	"time"
+)
+
+const maxStatActions = 15 // Keep stats for last X number of actions
+
+// Defines statistics kept by the agent
+type agentStats struct {
+	Actions []agentStatsAction
+	sync.Mutex
+}
+
+// Import information about an incoming action into agent statistics. If
+// accepted is true the action was executed by the agent, if false it means
+// the action was rejected (e.g., due to lack of signatures).
+func (s *agentStats) importAction(a mig.Action, accepted bool) error {
+	s.Lock()
+	defer s.Unlock()
+	ns := agentStatsAction{}
+	err := ns.importAction(a, accepted)
+	if err != nil {
+		return err
+	}
+	if len(s.Actions) == maxStatActions {
+		s.Actions = s.Actions[1:]
+	}
+	s.Actions = append(s.Actions, ns)
+	return nil
+}
+
+// Stats we keep for an action processed by the agent
+type agentStatsAction struct {
+	Time     string
+	Name     string
+	Accepted string
```

Comment:
Added modules

Investigator names is a bit trickier, that information is not sent with the action to the agents (appears this is only populated when actions are retrieved from the API). We'd have to pull this out of the key ring and there isn't a nice way to return that right now from the ACL verification code.